### PR TITLE
feat(machinery): allow base_url of OpenAI to be customized

### PR DESCRIPTION
**Pull request by max-wittig**
Allows usage of local LLMs or other providers

image

## Proposed changes

Adds a new field when setting up the OpenAI integration to allow the user to add a custom base_url

## Checklist

- [ ] Lint and unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added documentation to describe my feature.
- [ ] I have squashed my commits into logic units.
- [ ] I have described the changes in the commit messages.

### CODE REVIEW

**Comment by nijel** [comment]
Can you please add documentation which other providers or local LLM work with the OpenAI API and how to configure these?

File: weblate/machinery/openai.py [hunk 1]
```diff
@@ -64,7 +64,13 @@ class OpenAITranslation(BatchMachineTranslation):
...
     def __init__(self, settings=None) -> None:
         super().__init__(settings)
-        self.client = OpenAI(api_key=self.settings["key"], timeout=self.request_timeout)
+        kwargs = {
+            "api_key": self.settings["key"],
+            "timeout": self.request_timeout,
+        }
+        if base_url := self.settings.get("base_url"):
+            kwargs["base_url"] = base_url
+        self.client = OpenAI(**kwargs)
```
Comment (nijel) [inline weblate/machinery/openai.py:73]: Better use arguments directly for type checking:

```suggestion
        self.client = OpenAI(
            api_key=self.settings["key"], 
            timeout=self.request_timeout, 
            base_url=self.settings.get("base_url")
        )
```

File: weblate/machinery/forms.py [hunk 2]
```diff
@@ -310,3 +310,13 @@ class OpenAIMachineryForm(KeyMachineryForm):
         ),
         required=False,
     )
+    base_url = forms.URLField(
+        label=pgettext_lazy(
+            "Automatic suggestion service configuration",
+            "Translator Base URL",
```
Comment (nijel) [inline weblate/machinery/forms.py:316]: Perhaps OpenAI base URL?

Comment (max-wittig) [reply weblate/machinery/forms.py:316]:
```suggestion
            "OpenAI base URL",
```

File: weblate/machinery/forms.py [hunk 3]
```diff
@@ -310,3 +310,13 @@ class OpenAIMachineryForm(KeyMachineryForm):
...
     )
+    base_url = forms.URLField(
+        label=pgettext_lazy(
+            "Automatic suggestion service configuration",
+            "OpenAI base URL",
+        ),
+        widget=forms.Textarea,
+        help_text=gettext_lazy("Base URL of the OpenAI API"),
+        required=False,
+        default=None,
```
Comment (nijel) [inline weblate/machinery/forms.py:321]: TypeError: Field.__init__() got an unexpected keyword argument 'default'

See https://github.com/WeblateOrg/weblate/actions/runs/8800026958/job/24150336022?pr=11467

Seriously, has this been tested to wokr?

Comment (max-wittig) [reply weblate/machinery/forms.py:321]: @nijel No, I just pushed it to see the pipeline! Not ready for review! I should have marked it as a draft!

Thanks for the fast review though.

Comment (nijel) [reply weblate/machinery/forms.py:321]: I recommend testing locally first against real custom API endpoint. The model selection will need to be changed for custom LLMs as well.

**Comment by max-wittig** [comment]
E.g. I tried also using an API client. Same result:

image

image

**Comment by max-wittig** [comment]
While our model and our intended use-case does not work, I still would like to get it in for people that use more powerful models (as I already made the change). @nijel Could you give it another review?

File: weblate/machinery/openai.py (hunk 1 above)
Comment (nijel) [inline weblate/machinery/openai.py:73]: Make type checking possible:

```suggestion
        self.client = OpenAI(
            api_key=self.settings["key"], 
            timeout=self.request_timeout,
            base_url=self.settings.get("base_url") or None,
        )
```

File: weblate/machinery/forms.py [hunk 4]
```diff
@@ -278,8 +278,19 @@ class OpenAIMachineryForm(KeyMachineryForm):
         ("gpt-4", "GPT-4"),
         ("gpt-3.5-turbo-1106", "Updated GPT 3.5 Turbo"),
         ("gpt-3.5-turbo", "GPT-3.5 Turbo"),
+        ("custom", "Custom model"),
```
Comment (nijel) [inline weblate/machinery/forms.py:281]:
```suggestion
        ("custom", pettext_lazy("OpenAI model selection", "Custom model")),
```

File: weblate/machinery/forms.py [hunk 5]
```diff
@@ -288,6 +299,14 @@ class OpenAIMachineryForm(KeyMachineryForm):
         initial="auto",
         choices=MODEL_CHOICES,
     )
+    custom_model = forms.CharField(
+        label=pgettext_lazy(
+            "Custom model name",
```
Comment (nijel) [inline weblate/machinery/forms.py:304]:
```suggestion
            "OpenAI model selection",
```

File: weblate/machinery/forms.py [hunk 6]
```diff
@@ -288,6 +299,14 @@ class OpenAIMachineryForm(KeyMachineryForm):
...
+            "Custom model name",
+        ),
+        help_text=gettext_lazy("Only needed when model is set to 'Custom model'"),
+        required=False,
+    )
```
Comment (nijel) [inline weblate/machinery/forms.py:309]: This deserves additional form validation that it is non-blank when custom model choice is selected.

File: weblate/machinery/tests.py [hunk 7]
```diff
@@ -1489,6 +1489,7 @@ class OpenAITranslationTest(BaseMachineTranslationTest):
         "model": "auto",
         "persona": "",
         "style": "",
+        "base_url": "https://api.openai.com/v1",
```
Comment (nijel) [inline weblate/machinery/tests.py:1492]: Should not be needed:

```suggestion
```

On the other side, can you please add a separate test case to verify that `base_url` and custom model are actually being used?

**Review by nijel — COMMENTED** [review]
I've committed my suggestions, what needs to be addressed:

* `OpenAIMachineryForm` should have validation that custom model is provided only iff custom model is selected in the dropdown
* Test to verify that base URL change actually applies

File: weblate/machinery/forms.py [hunk 8]
```diff
@@ -278,8 +278,19 @@ class OpenAIMachineryForm(KeyMachineryForm):
         ("gpt-4", "GPT-4"),
         ("gpt-3.5-turbo-1106", "Updated GPT 3.5 Turbo"),
         ("gpt-3.5-turbo", "GPT-3.5 Turbo"),
+        ("custom", pettext_lazy("OpenAI model selection", "Custom model")),
```
Comment (nijel) [inline weblate/machinery/forms.py:281]:
```suggestion
        ("custom", pgettext_lazy("OpenAI model selection", "Custom model")),
```

**Comment by nijel** [comment]
I've implemented the missing bits and scheduled this for merge. Thanks for your contribution!

**Comment by max-wittig** [comment]
Thanks for taking care!

**Pull request merged**